share/eds/store.go

```diff
@@ -197,16 +200,42 @@ func (s *Store) Blockstore() bstore.Blockstore {
 	return s.bs
 }
 
-// CARBlockstore returns the IPFS blockstore that provides access to the IPLD blocks stored in an
-// individual CAR file.
-func (s *Store) CARBlockstore(ctx context.Context, dataHash []byte) (dagstore.ReadBlockstore, error) {
-	key := shard.KeyFromString(fmt.Sprintf("%X", dataHash))
+// CARBlockstore returns the IPFS blockstore and DAH that provides access to the IPLD blocks stored
+// in an individual CAR file.
+func (s *Store) CARBlockstore(
+	ctx context.Context,
+	dataHash share.DataHash,
+) (dagstore.ReadBlockstore, *share.Root, error) {
+	key := shard.KeyFromString(dataHash.String())
 	accessor, err := s.getAccessor(ctx, key)
 	if err != nil {
-		return nil, err
+		return nil, nil, fmt.Errorf("failed to get accessor: %w", err)
 	}
 
-	return accessor.bs, nil
+	reader, err := carv1.NewCarReader(accessor.sa)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to read DAH from car reader: %w", err)
+	}
+	dah := dahFromCARHeader(reader.Header)
+	err = accessor.sa.Close()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to close car reader after DAH retrieval: %w", err)
```

Comment:
We cannot close the accessor and use the bs afterward, as it closes the mmap internally.